src/dbnode/storage/index.go

```diff
@@ -2137,20 +2137,18 @@ func (i *nsIndex) CleanupDuplicateFileSets() error {
 	filesToDelete := make([]string, 0)
 	for _, segmentsOrderByVolumeIndexByVolumeType := range segmentsOrderByVolumeIndexByVolumeTypeAndBlockStart {
 		for _, segmentsOrderByVolumeIndex := range segmentsOrderByVolumeIndexByVolumeType {
-			shardTimeRangesCovered := result.NewShardTimeRanges()
-			currSegments := make([]fs.Segments, 0)
+			segmentsToKeep := make([]fs.Segments, 0)
 			for _, seg := range segmentsOrderByVolumeIndex {
-				if seg.ShardTimeRanges().IsSuperset(shardTimeRangesCovered) {
-					// Mark dupe segments for deletion.
-					for _, currSeg := range currSegments {
-						filesToDelete = append(filesToDelete, currSeg.AbsoluteFilePaths()...)
+				for len(segmentsToKeep) > 0 {
+					idx := len(segmentsToKeep) - 1
+					if previous := segmentsToKeep[idx]; seg.ShardTimeRanges().IsSuperset(previous.ShardTimeRanges()) {
```

Comment:
Do we need to extract the active shards to compare here? 

Shouldn't `IsSuperset()` return `false` if the previous shard time ranges contain a shard missing from the current ranges? Like when we remove a shard.

```
// Check if other ranges has any shards we do not have.
for shard := range other.Iter() {
        if _, ok := r.Get(shard); !ok {
                return false
        }
}
```

Reply:
Not quite sure I get what you mean. Do you say that only shards from the most recent volume should be taken into account (ignoring any extra shards that might be there in older volumes)? E.g. if I have volumes with shards:
```
volume index: shard list
0: 0, 1, 2, 3
1: 0, 1, 2
2: 0, 1
```
I should only check shards {0, 1} and their time ranges when comparing volumes? And  shards {2, 3} should be skipped, since those only appear in obsolete volumes?

Reply:
@notbdu could you review these changes?